### DISCUSSION TITLE
Private attributes and multiple inheritance.

### DISCUSSION THREAD

**Discussion by andregerbaulet** [pydantic/pydantic#2740] [Question]
# TLDR
- Want subclass to inherit from two or more _pydantic_ "BaseModel" parent classes.
- Private attributes cause "TypeError: multiple bases have instance lay-out conflict".
- Is multiple inheritance supported by _pydantic_ and if so, what is the correct approach?

# Private attributes and multiple inheritance.

Hi, I'm having some issues with multiple inheritance and _pydantic_. Not entirely sure if this is bug or error by me, but it seems like odd behaviour. What i want is to create a subclass ("Child" in code below) that inherits from two or more parent classes which have their own field and attributes and the validation and parsing that comes with _pydantic_. My naïve solution was the following.

```py
from pydantic import BaseModel, PrivateAttr

class Parent(BaseModel):
    public_name: str = 'Bruce Wayne'
    _secret_name: str = PrivateAttr()


class AnotherParent(BaseModel):
    public_number: int = 21
    _secret_number: int = PrivateAttr()


class Child(Parent, AnotherParent):
    public_friend: str = 'Alfred'
    _secret_friend: str = PrivateAttr()

    def __init__(self, **kwargs):
        super().__init__(**kwargs)
        self._secret_name = 'Batman'
        self._secret_number = 42
        self._secret_friend = 'The Joker'


if __name__ == '__main__':
    child = Child()
    print(child.dict())
```

Running this yields the following traceback:

```
Traceback (most recent call last):
  File "python_test.py", line 11, in <module>
    class Child(Parent, AnotherParent):
  File "pydantic/main.py", line 324, in pydantic.main.ModelMetaclass.__new__
  File "/Users/andregerbaulet/opt/anaconda3/envs/enfin/lib/python3.8/abc.py", line 85, in __new__
    cls = super().__new__(mcls, name, bases, namespace, **kwargs)
TypeError: multiple bases have instance lay-out conflict
```

This in itself might not be unusual as both "Parent" and "AnotherParent" inherits from "BaseModel" which perhaps causes some conflicts. However, just removing the private attributes of "AnotherParent" makes it work as expected.

```py
from pydantic import BaseModel, PrivateAttr

class Parent(BaseModel):
    public_name: str = 'Bruce Wayne'
    _secret_name: str = PrivateAttr()


class AnotherParent(BaseModel):
    public_number: int = 21
    # _secret_number: int = PrivateAttr()     <-- This has been commented out


class Child(Parent, AnotherParent):
    public_friend: str = 'Alfred'
    _secret_friend: str = PrivateAttr()

    def __init__(self, **kwargs):
        super().__init__(**kwargs)
        self._secret_name = 'Batman'
        # self._secret_number = 42            <-- This has been commented out
        self._secret_friend = 'The Joker'


if __name__ == '__main__':
    child = Child()
    print(child.dict())
```

Which yields:

```
{'public_number': 21, 'public_name': 'Bruce Wayne', 'public_friend': 'Alfred'}
```

My questions are:
- Is multiple inheritance supported by _pydantic_?
- If so, what is the correct way to accomplish this?

Python version == 3.8.8
_pydantic_ version == 1.7.3

**Comment by PrettyWood** [comment]
Ah the famous sloted classes multiple inheritance problem! It does ring a bell 😆 
I can maybe have a look tonight I'm on my phone right now.

**Reply by andregerbaulet** [reply]
Thank you, I would very much appreciate it! Have been stuck on this for quite some time now... :/
